src/envs/tools.py

```diff
@@ -0,0 +1,646 @@
+                       objects: Sequence[Object], params: Array) -> Action:
+        del memory  # unused
+        assert not params
+        item = objects[1]
+        assert self._is_item(item)
+        pose_x = state.get(item, "pose_x")
+        pose_y = state.get(item, "pose_y")
+        arr = np.array([pose_x, pose_y, 0.0, 0.0], dtype=np.float32)
+        return Action(arr)
+
+    def _get_closest_item_or_tool(self, state: State, x: float,
+                                  y: float) -> Optional[Object]:
+        closest_obj = None
+        closest_dist = float("inf")
+        for obj in state:
+            if obj == self._robot:
+                continue
+            if obj.type == self._contraption_type:
+                continue
+            x_dist = abs(state.get(obj, "pose_x") - x)
+            y_dist = abs(state.get(obj, "pose_y") - y)
+            if x_dist > self.close_thresh or y_dist > self.close_thresh:
+                continue
+            dist = x_dist + y_dist
+            if dist < closest_dist:
+                closest_dist = dist
+                closest_obj = obj
+        return closest_obj
+
+    def _get_held_item_or_tool(self, state: State) -> Optional[Object]:
+        held_obj = None
+        for obj in state:
+            if obj == self._robot:
+                continue
+            if obj.type == self._contraption_type:
+                continue
+            if state.get(obj, "is_held") > 0.5:
+                assert held_obj is None
+                held_obj = obj
+        return held_obj
+
+    def _get_contraption_pose_is_on(self, state: State, x: float,
+                                    y: float) -> Optional[Object]:
+        for obj in state:
+            if obj.type != self._contraption_type:
+                continue
+            if self._is_pose_on_contraption(state, x, y, obj):
+                return obj
+        return None
+
+    @staticmethod
+    def _is_pose_on_contraption(state: State, x: float, y: float,
+                                contraption: Object) -> bool:
+        pose_lx = state.get(contraption, "pose_lx")
+        pose_ly = state.get(contraption, "pose_ly")
+        pose_ux = state.get(contraption, "pose_ux")
+        pose_uy = state.get(contraption, "pose_uy")
+        return pose_lx < x < pose_ux and pose_ly < y < pose_uy
+
+    def _is_tool(self, obj: Object) -> bool:
+        return obj.type in (self._screwdriver_type, self._hammer_type,
+                            self._wrench_type)
+
+    def _is_item(self, obj: Object) -> bool:
+        return obj.type in (self._screw_type, self._nail_type, self._bolt_type)
+
+    def _is_screwdriver_or_hammer(self, obj: Object) -> bool:
+        return obj.type in (self._screwdriver_type, self._hammer_type)
+
+    def _get_best_screwdriver_or_none(self, state: State,
+                                      screw: Object) -> Optional[Object]:
+        """Use the shape of the given screw to figure out the best graspable
```

Comment:
no action required, just want to say this so i remember: it's counterintuitive that whether a screw is fastenable by hand depends on properties of the other screwdrivers

Reply:
yeah it's just a way of ensuring we don't have tons of low-level features lol. you can think of it like "if i have the right screwdriver for the job, i'll use that, but otherwise i'll use my hands like an ape"